### ISSUE TITLE
period = x.shape[1] // top_list   top_list里边是frequency的 index,请问这里用时序长度/index，为何得到的是周期？？？如果/frequency似乎更容易理解，期待回复，非常感谢！

### ISSUE THREAD

**Issue by hnczj** [thuml/Time-Series-Library#210]
def FFT_for_Period(x, k=2):#x torch.Size([2, 31, 32]) #31?????############

# [B, T, C]

xf = torch.fft.rfft(x, dim=1)#torch.Size([2, 16, 32])

# find period by amplitudes

frequency_list = abs(xf).mean(0).mean(-1)#abs measures how strong the signal is

##both x and y are averaged along batch_size and along (x and y)

frequency_list[0] = 0 # common practice

_, top_list = torch.topk(frequency_list, k)#return values indexes

#_, top_list = torch.topk(frequency_list[1:], k, largest=False)#, largest=False

#pdb.set_trace()

top_list = top_list.detach().cpu().numpy()

# # Get all indices and shuffle them

# all_indices = np.arange(len(frequency_list[1:]))

# np.random.shuffle(all_indices)

# # Select the first k shuffled indices

# top_list = all_indices[:k]

##用days除以頻率，確定每個k有幾個週期

period = x.shape[1] // top_list# higher index means higher frequency

#pdb.set_trace()

return period, abs(xf).mean(-1)[:, top_list]                                                                                                                 period = x.shape[1] // top_list                                                                                                                      top_list里边是frequency的 index,请问这里用时序长度/index，为何得到的是周期？？？如果/frequency似乎更容易理解，期待回复，非常感谢！                                             read the whole code, top_list here is index of frequency , why here x.shape[1] //top_list = period, if use frequency, it seems meaningful, but using index of frequency seems meaningless

**Comment by hnczj** [comment]
this is gpt4's answer, is this explanation ok?                                                                                                  The line of code `period = x.shape[1] // top_list` is calculating the period of each of the highest `k` frequencies found in the time series data.

In the context of Fourier analysis, frequency and period are inversely related. The higher the frequency, the shorter the period, and vice versa. Since `top_list` contains the indices of the top `k` frequencies, these indices are used as a proxy for the frequencies themselves.

The reasoning behind this is that the output of the FFT function in PyTorch is ordered by increasing frequency. That is, the first component (index 0) corresponds to the lowest frequency (DC or 0 frequency), the second component (index 1) corresponds to the next higher frequency, and so on. Therefore, the index of each component in the FFT output can be seen as a representation of its frequency.

Therefore, `x.shape[1] // top_list` calculates the period corresponding to each of the top `k` frequencies. `x.shape[1]` corresponds to the total length of the time series data (or the total number of time steps), and `//` is the integer division operator in Python. Thus, the line of code is dividing the total length of the time series by each of the top `k` frequencies (represented by their indices) to calculate their corresponding periods.

However, this is a somewhat rough approximation, as it assumes a linear relationship between the index in the FFT output and the actual frequency, which may not always be the case, especially for longer time series or those with non-uniform sampling rates. But in many cases, especially for simple analysis or initial exploratory work, this approximation can still provide useful insights about the periodic patterns in the time series data.

**Comment by hnczj** [comment]
but in my example,frequency_list =  tensor([0.0000, 7.2521, 2.8845, 2.9960, 5.1468, 2.2075, 4.2677, 3.7666, 3.1520,
        6.1255, 4.6509, 4.9537, 3.1828, 7.7338, 7.4764, 3.3870],
       device='cuda:0', grad_fn=<CopySlices>)           then top_list = [2,2]

**Comment by wuhaixu2016** [comment]
您好，请详细读一下torch.fft的含义：https://pytorch.org/docs/stable/generated/torch.fft.fft.html#torch.fft.fft
fft的结果只是说明了每个频率的强度，index表示频率的值。

**Comment by hnczj** [comment]
您好，来信已收到，多谢！

**Comment by hnczj** [comment]
非常感谢您的解释！我从中学到了很多！根据我从OpenAI的GPT-4获取的信息，我想补充一些关于torch.fft.fft函数的信息...这个函数是PyTorch库中用于计算一维离散傅立叶变换（Discrete Fourier Transform，简称DFT）的。它接收一个输入张量，返回一个与输入同形状的复数张量，这个复数张量中的每个元素都表示输入中对应频率的振幅和相位。所以，`torch.fft.fft`函数的输出实际上是一种频域表示，反映了输入信号在各个频率上的强度。在FFT的输出中，"频率"是由其对应的索引值（index）表示的，而每个频率的"强度"则是由该频率对应的复数值的模来表示的。然而，FFT输出的索引与实际频率之间的关系并非一对一的。在FFT输出中，索引值从0开始，到`n/2`（`n`为FFT的大小或者输入信号的长度）时，对应的频率从0Hz增加到采样频率的一半。然后，从`n/2`到`n`的索引值，对应的频率从负的采样频率的一半增加到接近0Hz。因此，虽然可以将FFT输出的索引看作是表示频率的值，但这种表示方式是基于特定的FFT大小和采样频率的，并且对于大于`n/2`的索引值，其实是表示负频率。

FFT输出的索引可以看作是频率的表示，而具体对应到何种频率则取决于数据的采样率和FFT的大小。同时，我也了解到，通过计算 `abs(xf).mean(0).mean(-1)` 得到的是所有频率分量强度的整体度量，而不是特定的频率或索引。

关于如何确定信号中最强的频率，我认为这是一个非常值得深入研究的话题，我会继续思考和探索。非常期待和大家一起讨论和交流。

再次感谢您的回答，祝您一切顺利。
